Add Phi-1 and Phi-1_5

susnato opened the pull request:
# What does this PR do?

<!--
Congratulations! You've made it this far! You're not quite done yet though.

Once merged, your PR is going to appear in the release notes with the title you set, so make sure it's a great title that fully reflects the extent of your awesome contribution.

Then, please replace this with a description of the change and which issue is fixed (if applicable). Please also include relevant motivation and context. List any dependencies (if any) that are required for this change.

Once you're done, someone will review your PR shortly (see the section "Who can review?" below to tag some potential reviewers). They may suggest changes to make the code even better. If no one reviewed your PR after a week has passed, don't hesitate to post a new comment @-mentioning the same persons---sometimes notifications get lost.
-->

<!-- Remove if not applicable -->

Fixes #26110 


## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?


[Colab Link](https://colab.research.google.com/drive/1rY-T36vqG9b-0IM3IJq741zBlcSxSit1?usp=sharing) to verify logits.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc : @xenova 

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

Models:

- text models: @ArthurZucker and @younesbelkada
- vision models: @amyeroberts
- speech models: @sanchit-gandhi
- graph models: @clefourrier

Library:

- flax: @sanchit-gandhi
- generate: @gante
- pipelines: @Narsil
- tensorflow: @gante and @Rocketknight1
- tokenizers: @ArthurZucker
- trainer: @muellerz and @pacman100

Integrations:

- deepspeed: HF Trainer/Accelerate: @pacman100
- ray/raytune: @richardliaw, @amogkam
- Big Model Inference: @SunMarc
- quantization (bitsandbytes, autogpt): @SunMarc and @younesbelkada

Documentation: @stevhliu and @MKhalusova

HF projects:

- accelerate: [different repo](https://github.com/huggingface/accelerate)
- datasets: [different repo](https://github.com/huggingface/datasets)
- diffusers: [different repo](https://github.com/huggingface/diffusers)
- rust tokenizers: [different repo](https://github.com/huggingface/tokenizers)

Maintained examples (not research project or legacy):

- Flax: @sanchit-gandhi
- PyTorch: See Models above and tag the person corresponding to the modality of the example.
- TensorFlow: @Rocketknight1

 -->

susnato committed:
only dir not even init

susnato committed:
init

susnato committed:
tokenizer removed and reference of codegen added

LysandreJik commented:
Hey @susnato, thanks for opening your PR! Let us know if we can help in any way.

Also cc @gugarosa and @sgunasekar as you've been doing impressive maintenance work on your models :hugs:

susnato committed:
modeling file updated a lot remaining app_rotary_emb

susnato committed:
conversion script done

susnato committed:
conversion script fixed, a lot of factoring done and most tests pass

susnato committed:
added token_clf and extractive_QA_head

susnato committed:
integration tests pass

susnato committed:
flash attn tests pass!

susnato committed:
config done

susnato committed:
more docs in modeling file

susnato committed:
some style fix

susnato committed:
style and others

susnato committed:
doc test error fix

susnato committed:
more doc fix

susnato commented:
Hi @LysandreJik, this PR is ready for review.
Please refer to this [colab notebook](https://colab.research.google.com/drive/1rY-T36vqG9b-0IM3IJq741zBlcSxSit1?usp=sharing) to check the logits matching for `Phi-1` and `Phi-1.5`.

BTW the test - `test_left_padding_compatibility` seems to be failing here, but it is passing on my local machine. Do you know what could be causing this discrepancy?

LysandreJik commented:
Nice, thanks a lot! Pinging @ArthurZucker and @Rocketknight1 to take a look :)

ArthurZucker commented on src/transformers/models/phi/modeling_phi.py:
```diff
@@ -0,0 +1,1339 @@
+        return hidden_states
+
+
+class PhiAttention(nn.Module):
```
You should have a look at `PersimmonAtttention` ! It also has a rotary embedding applied to only a portion of the input.

susnato replied:
The attention class for `Phi` is almost same for `Persimmon` except the slight difference in `_split_heads` method. 
I have added `# Copied from` statements to all methods(init, forward, _init_rope) in the attention class except the _split_heads.

ArthurZucker commented on src/transformers/models/phi/modeling_phi.py:
```diff
@@ -0,0 +1,1339 @@
+        )
+
+
+def apply_rotary_pos_emb(q, k, cos, sin, position_ids, rotary_dim):
```
similar to the Persimmon model it seems. Only apply it on `rotary_dim` portion of the input

susnato replied:
yes, updates it in the recent commit.

ArthurZucker reviewed (commented):
Hey! Seems to me that the modeling is the same as `Persimmon` with just `kq_layer_norm = False` in the config and a different `config.partial_rotary_factor`. 
WDYT?

SinclairCoder commented:
Will this PR support the training for phi models?

susnato committed:
some attention fixes

susnato committed:
most fixes

susnato committed:
style and other fixes

susnato commented:
Hi @SinclairCoder, yes it will.

susnato committed:
docs fix and config

susnato commented:
Hello @ArthurZucker, thanks for the suggestions and apologies for the delay.  

I have changed the code quite a bit according to the `persimmon`. For the `PhiAttention`, there is slight difference between the `_split_heads` methods of phi and persimmon so I could not just add `# Copied from` across the whole class, instead I have copied the relevant methods seperately.
In addition to that please note that there are subtle differences in the `PhiDecoderLayer` and `PhiModel` too.

Please review it and let me know if you are ok with the changes or not.

susnato committed:
doc fix

ArthurZucker commented on src/transformers/models/phi/modeling_phi.py:
```diff
@@ -0,0 +1,1201 @@
+"""
+
+
+@add_start_docstrings(
+    "The bare Phi Model outputting raw hidden-states without any specific head on top.",
+    PHI_START_DOCSTRING,
+)
+class PhiModel(PhiPreTrainedModel):
```
expecting to see a lot more copied from here

susnato replied:
Done

ArthurZucker reviewed (commented):
Thanks for updating ! A few nits but should be close to done

susnato commented on src/transformers/models/phi/modeling_phi.py:
```diff
@@ -0,0 +1,1201 @@
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+        )
+
+        sequence_output = outputs[0]
+
+        logits = self.qa_outputs(sequence_output)
+        start_logits, end_logits = logits.split(1, dim=-1)
+        start_logits = start_logits.squeeze(-1).contiguous()
+        end_logits = end_logits.squeeze(-1).contiguous()
+
+        total_loss = None
+        if start_positions is not None and end_positions is not None:
+            # If we are on multi-GPU, split add a dimension
+            if len(start_positions.size()) > 1:
+                start_positions = start_positions.squeeze(-1)
+            if len(end_positions.size()) > 1:
+                end_positions = end_positions.squeeze(-1)
+            # sometimes the start/end positions are outside our model inputs, we ignore these terms
+            ignored_index = start_logits.size(1)
+            start_positions = start_positions.clamp(0, ignored_index)
+            end_positions = end_positions.clamp(0, ignored_index)
+
+            loss_fct = CrossEntropyLoss(ignore_index=ignored_index)
+            start_loss = loss_fct(start_logits, start_positions)
+            end_loss = loss_fct(end_logits, end_positions)
+            total_loss = (start_loss + end_loss) / 2
+
+        if not return_dict:
+            output = (start_logits, end_logits) + outputs[2:]
+            return ((total_loss,) + output) if total_loss is not None else output
+
+        return QuestionAnsweringModelOutput(
+            loss=total_loss,
+            start_logits=start_logits,
+            end_logits=end_logits,
+            hidden_states=outputs.hidden_states,
+            attentions=outputs.attentions,
+        )
```
Yes there are no checkpoints.

susnato commented on src/transformers/models/phi/modeling_phi.py:
```diff
@@ -0,0 +1,1201 @@
+            value: [batch_size, seq_length, num_heads, head_dim]
+        """
+        batch_size, seq_length, three_times_hidden_size = fused_qkv.shape
+        fused_qkv = fused_qkv.view(batch_size, seq_length, 3, self.num_heads, self.head_dim)
```
> so the subtle difference is the shape that is (3, num_heads) instead of (num_heads, 3)?

yes

susnato replied:
Done

susnato committed:
some comments

susnato committed:
conversion script updated

susnato committed:
conversion script updated

susnato committed:
Revert "conversion script updated"

This reverts commit e92378c54084ec0747041b113083d1746ecb6c7f.

susnato committed:
final comments

susnato committed:
add Phi to language_modeling.md

susnato commented:
Hi @ArthurZucker ,  I have worked on the comments and now `PhiAttention` and `PhiModel` are same as their Persimmon counterparts. 

Please review it and let me know if this works with you.

ArthurZucker commented on docs/source/en/model_doc/phi.md:
```diff
@@ -0,0 +1,137 @@
+The Phi-1.5 model was proposed in [Textbooks Are All You Need II: phi-1.5 technical report](https://arxiv.org/abs/2309.05463) by Yuanzhi Li, Sébastien Bubeck, Ronen Eldan, Allie Del Giorno, Suriya Gunasekar and Yin Tat Lee.
+
+### Summary
+In Phi-1 and Phi-1.5 papers, the authors showed how important the quality of the data is in training relative to the model size.
+They selected high quality "textbook" data alongside with synthetically generated data for training their small sized Transformer
+based model Phi-1 with 1.3B parameters. Despite this small scale, phi-1 attains pass@1 accuracy 50.6% on HumanEval and 55.5% on MBPP.
+They follow the same strategy for Phi-1.5 and created another 1.3B parameter model with performance on natural language tasks comparable 
+to models 5x larger, and surpassing most non-frontier LLMs. Phi-1.5 exhibits many of the traits of much larger LLMs such as the ability 
+to “think step by step” or perform some rudimentary in-context learning.
+With these two experiments the authors successfully showed the huge impact of quality of training data when training machine learning models.
```
🔥

ArthurZucker commented on src/transformers/models/phi/configuration_phi.py:
````diff
@@ -0,0 +1,175 @@
+    >>> # Accessing the model configuration
+    >>> configuration = model.config
+    ```"""
+    model_type = "phi"
````
```suggestion
    model_type = "phi"
```
We are gonna have the same problem as with Faclon, the online model type is `"model_type": "mixformer-sequential",`.
For this, I think we'll have to either keep it or find a workaround! (Otherwise TGI could break)
but that will be the last thing to solve!

susnato replied:
Can't we just change the `model_type` to `phi` when we update the weights there?

ArthurZucker replied:
No let's just use the one they have online that will be the simplest!

susnato replied:
Hi @ArthurZucker, I changed the model_type from `"phi"` to `"mixformer-sequential"` but I am getting the `test_load_with_mismatched_shapes` error, I tried to change the mappings in `models/auto` but still getting the error...

susnato replied:
I have changed `model_type = "phi" -> "mixformer-sequential"` and added a line in the test modelling file `test_mismatched_shapes = False` which skips the failed test for now. 
Later after we transfer the weights and update the online config we can revert back to `model_type = "phi"` and remove the line.

WDYT?

ArthurZucker replied:
No the error is here to tell you that you need to update the `config_mapin` meaning you need to make sure that `configuration_auto` also has the correct name! 😉 
Let's add a test to make sure that if we load a model from the current revision online with this code , the generated outputs are correct! (using the code_revision and revision flags!)

susnato replied:
Hey I did change the `CONFIG_MAPPING_NAMES` from `("phi", "PhiConfig")->("mixformer-sequential", "PhiConfig")` But the error was still there :), 
Is there anything more I have to change?

susnato replied:
Also we can't load either phi-1 or phi-1.5 from the current online version because the weights(namings) won't match :(

susnato replied:
Please see [this comment](https://github.com/huggingface/transformers/pull/26170#issuecomment-1778720835).

ArthurZucker commented on src/transformers/models/phi/modeling_phi.py:
```diff
@@ -0,0 +1,1117 @@
+        config: PhiConfig
+    """
+
+    # Copied from transformers.models.llama.modeling_llama.LlamaModel.__init__ with LLAMA->PHI,Llama->Phi,PhiRMSNorm->nn.LayerNorm,norm->final_layernorm,rms_final_layernorm_eps->layer_norm_eps
```
its okay to leave that one out 🤣

susnato replied:
Hey we need to keep this line otherwise I am getting repo-consistency error :)

susnato replied:
Ok

ArthurZucker reviewed (commented):
LGTM, I want to approve but we're gonna have a BC compatibility issue with the model type in the config. That's my only concern!

ArthurZucker commented:
(also make sure the CI are green 🤗 )

susnato committed:
edit phi.md file

susnato committed:
rebase and fix

susnato committed:
removed phi-1.5 example

susnato commented on docs/source/en/model_doc/phi.md:
```diff
@@ -0,0 +1,137 @@
+
+This model was contributed by [Susnato Dhar](https://huggingface.co/susnato).
+The original code for Phi-1 and Phi-1.5 can be found [here](https://huggingface.co/microsoft/phi-1/blob/main/modeling_mixformer_sequential.py) and [here](https://huggingface.co/microsoft/phi-1_5/blob/main/modeling_mixformer_sequential.py) respectively.
+
```
Done

susnato commented:
Hello @ArthurZucker, I have pushed the changes and The CI is green now!

Regarding the BC compatibility issue - 
If it's ok with you, I would like to tackle that step by step after the PR is merged :smiley: .

1. Let's work together to make this PR merged.
2. After it is merged I will create a PR on the huggingface hub to upload both of the weights.
3. After that I will update the checkpoints, demo etc with separate PR just like we did for pop2piano. We can then tackle the compatibility issue.

Please let me know if this works with you :).

ArthurZucker commented:
For sure, then we should use the same model_type as them, this way we won't have any problems and should not affect anyone!

ArthurZucker commented:
(The issue is taht other libraries like TGI use the model_type and otherwise we need to wait for them)

susnato committed:
changed model_type from 'phi'->'mixformer-sequential'

susnato commented:
Hi @ArthurZucker, I have updated the model_type as theirs(`"mixformer-sequential"`) and mentioned the error I am getting(after the updation) and temporary fix [here](https://github.com/huggingface/transformers/pull/26170#discussion_r1370124238).

susnato committed:
small change

susnato committed:
small change

susnato committed:
revert \small change

ArthurZucker commented:
I am also super down to have the correct model type but we know that's gonna have issue. Last time we had to revert the falcon PR 😓 We don't have the perfect solution yet, I can work on this before we actually push the checkpoints and the config if you are okay to wait? Otherwise we can use their model type for now and later on once I find the correct way to approach this we'll update! 😉

susnato commented:
> I am also super down to have the correct model type but we know that's gonna have issue. Last time we had to revert the falcon PR 😓 We don't have the perfect solution yet, I can work on this before we actually push the checkpoints and the config if you are okay to wait?

If you think that's relatively better solution, then I will wait. 

But then could you please explain me the whole process a little bit?

Also I have a question - after the PR is merged, we will need to delete the modelling files on the HuggingFace Hub right? So that everyone loads the model from the library only?

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26170). All of your documentation changes will be reflected on that endpoint.

bnicholl commented:
when importing phi with transformers == 4.34.1 in the way the below docs describe. `from transformers import PhiModel`
[](https://moon-ci-docs.huggingface.co/docs/transformers/pr_26170/en/model_doc/phi)

I get an error: ImportError: cannot import name 'PhiForCausalLM'

ArthurZucker commented:
Hey @susnato, we'll probably merge this PR, then open a PR to the repo on the hub with the new checkpoints and a deletion of the code, making sure checking out this revision works, then merging once the release happens (next week). I'll let @Rocketknight1 help you with this as he went thought that for Falcon! 

Regarding this PR, let's merge it but with the correct model_type (so the one they have in their config)  for now. I didn't have time to dive into fixing and will be off for a week so will tackle after that!

susnato commented:
Thanks for the reply @ArthurZucker !

susnato commented:
Hey @ArthurZucker, I tried to change to `mixformer-sequential` for `MODEL_NAMES_MAPPING` and `CONFIG_MAPPING_NAMES` but it seems that I need to change the whole directory from `phi` to `mixformer-sequential` in order to make it work.
 

Otherwise I believe that we can also add  `("mixformer-sequential", "phi")` to `SPECIAL_MODEL_TYPE_TO_MODULE_NAME` so that every `mixformer-sequential` will get converted to `phi`.

WDYT?

ArthurZucker commented:
Mmm yeah something like `HUB_MODEL_TYPE_TO_TRANSFORMERS_MODEL_TYPE` will have to be created, but it's a bit more involved we need to test and make sure we handle edge case

ArthurZucker commented:
I'll try to do this today 👍🏻

susnato commented:
So what do we do now? Do we merge with model_type as phi? or keep the mixformer-sequential(but override the case as False)?

Rocketknight1 commented:
Just replying here to say that yes, I'm available to help when you're ready to start porting checkpoints!

susnato commented:
Hello @Rocketknight1, thanks for the reply!

I have already ported the checkpoints and saved here - [phi-1](https://huggingface.co/susnato/phi-1_dev) and [phi-1.5](https://huggingface.co/susnato/phi-1_5_dev)

The main [issue](https://github.com/huggingface/transformers/pull/26170#pullrequestreview-1692913227) that @ArthurZucker was talking about was the difference between the `model_type`, the [online config file, on the hub](https://huggingface.co/microsoft/phi-1/blob/371fd519ab41b5bc2e96838936c2349dc49656db/config.json#L14) has model_type as `mixformer-sequential` but here we have it as `phi`. This won't matter if we change the model_type to phi on the Hub.
Also I believe that we will need to delete the modelling files on the HuggingFace Hub right?

So, how should I approach this? should I first open a PR to transfer the weights and config to the original repo on Hub or this PR needs to be merged first then the weights transfer? WDYT?

Rocketknight1 commented:
@susnato This PR will need to be merged first, because otherwise the updated repos won't work for anyone else.

After that, the way we did it was to make a PR to each checkpoint repo, and do the following:

1) Copy the `configuration_phi.py` and `modeling_phi.py` files from this PR into the repo, overwriting the old ones.
2) Ensure the files don't use relative imports (e.g. replace `from ..modeling_utils import` with `from transformers.modeling_utils import`
3) Set the `architectures` and `auto_map` keys in `config.json` to point to the new classes.
4) Set the `model_type` to the new type (e.g. `phi` in this case)

This way, users who have updated to the latest version of `transformers` can use `trust_remote_code=False` and get the library code, but users who haven't can continue with their old workflows.

Rocketknight1 commented:
@susnato we should use "phi"

susnato committed:
changed mixformer-sequential->phi

susnato commented:
@Rocketknight1, Okk. Then lets merge this first.

If you don't mind could you please review this? :sweat_smile:

susnato committed:
small change

susnato committed:
added phi-1.5 example instead of phi-1

susnato committed:
doc test might pass now

Rocketknight1 commented:
@susnato will do! Please do the rebase anyway though, lol. I think we'll get lucky, and if we don't then it's better to find out here then after we merge!

Rocketknight1 commented on src/transformers/models/phi/modeling_phi.py:
```diff
@@ -0,0 +1,1116 @@
+
+
+# Copied from transformers.models.bart.modeling_bart._expand_mask
+def _expand_mask(mask: torch.Tensor, dtype: torch.dtype, tgt_len: Optional[int] = None):
+    """
+    Expands attention_mask from `[bsz, seq_len]` to `[bsz, 1, tgt_seq_len, src_seq_len]`.
+    """
+    bsz, src_len = mask.size()
+    tgt_len = tgt_len if tgt_len is not None else src_len
+
+    expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)
+
+    inverted_mask = 1.0 - expanded_mask
+
+    return inverted_mask.masked_fill(inverted_mask.to(torch.bool), torch.finfo(dtype).min)
```
This isn't a huge problem, but for new models we prefer to use [AttentionMaskConverter._expand_mask()](https://github.com/huggingface/transformers/blob/main/src/transformers/modeling_attn_mask_utils.py#L150) for this! It's exactly the same code, it just avoids everyone copying this one function from BART.

susnato replied:
Ok

susnato replied:
Done

Rocketknight1 reviewed (approved):
Given the similarities with `Persimmon`, this looks like a pretty clean port without a lot of deep code to review, so I'm happy to accept it!

I'd like two things before we merge, though:

1) The rebase (sorry!)
2) To double-check with Arthur and the TGI team about the model name. I'm 99% sure `phi` is acceptable here, but I want to confirm all of that before we merge. I'll get this done by tomorrow!

susnato committed:
rebase and small change

susnato committed:
added the dropout layer

susnato committed:
more fixes

susnato commented:
Hello @Rocketknight1, I have rebased and did necessary changes.

Please note that I have also added an [dropout layer](https://github.com/susnato/transformers/blob/26cceddd5504d5f58d899469483a0245f082f0ee/src/transformers/models/phi/modeling_phi.py#L610) after the `input_embeds` which was missing in the previous commits.

Rocketknight1 commented:
Got it! Pinging either @arthurzucker or @amyeroberts for core maintainer final review. I checked with the TGI team and they said it would be fine to do the port and change the model type to `phi`.

amyeroberts commented on docs/source/en/model_doc/phi.md:
```diff
@@ -0,0 +1,120 @@
+>>> generated_output = model.generate(**tokens, use_cache=True, max_new_tokens=10)
+
+>>> tokenizer.batch_decode(generated_output)[0]
+'If I were an AI that had just achieved a breakthrough in machine learning, I would be thrilled'
```
So would I 🥲

susnato replied:
But it might have different motives!! :laughing:

amyeroberts commented on docs/source/en/model_doc/phi.md:
There was a recent [push to main](https://github.com/huggingface/transformers/commit/5964f820db1568d26298b37dea9db328185c7f7c) which included standardisation of our model doc page structures and section titles. Could you update this to reflect those e.g. using `Usage examples`, and putting tips inside an embedded `<Tip>` section

amyeroberts commented on src/transformers/models/phi/modeling_phi.py:
```diff
@@ -0,0 +1,1072 @@
+        >>> model = PhiForCausalLM.from_pretrained("susnato/phi-1_dev")
+        >>> tokenizer = AutoTokenizer.from_pretrained("susnato/phi-1_dev")
+
+        >>> prompt = "Hey, are you conscious? Can you talk to me?"
```
Let's avoid example where we talk about consciousness for models

bnicholl replied:
In the above example we use 
`model = PhiForCausalLM.from_pretrained("susnato/phi-1_dev")`, where PhiForCausalLM gives an import error.

In the official docs we're using 
`model = AutoModelForCausalLM.from_pretrained("microsoft/phi-1_5", trust_remote_code=True)`.

Are both of these functions of loading the model the same? I only ask because I know with previous versions, you could not fine tune the model, so I wanna make sure fine tuning is now corrected using both function calls.

susnato replied:
Hi @bnicholl, if you want to use `PhiForCausalLM` right now you need to checkout my branch (`susnato:phi`) and then install from there, otherwise wait for this PR to get merged. :)

> Are both of these functions of loading the model the same? I only ask because I know with previous versions, you could not fine tune the model, so I wanna make sure fine tuning is now corrected using both function calls.

Once this PR is merged it will load the same model and yes you can fine-tune like any other models from this library!

amyeroberts reviewed (commented):
Awesome work - thanks for adding this model! 

Just some nits on documentation and examples. Once updated we'll be good to go!

susnato committed:
modified .md file

susnato commented:
Hi @amyeroberts, I have reworked the model doc page. Let me know if this is okay to you.

amyeroberts reviewed (approved):
Thanks again for adding and iterating!

susnato commented:
Looks like the CI error is unrelated to this PR. Should I rebase and push again? @amyeroberts

amyeroberts commented:
@susnato Yes please! Sorry - there were a few unexpected error because of new packages releases and incompatibilities. These should now be resolved on main.

susnato commented:
Hi, @amyeroberts I just rebased but the failing test seems to be unrelated.

amyeroberts commented:
@susnato Indeed. Really sorry about this, we're trying to resolve now. It seems there's some flakiness in accelerate tests - which means even if we _think_ we've fixed them another rears its head. I'll let you know once the latest patch is merged. Fingers crossed that will be the last rebase you'll have to do.

susnato commented:
No problem @amyeroberts, just ping me when it's patched and I will rebase then.

amyeroberts commented:
OK, third time lucky. Apologies for the delay in getting this merged and thank you for your patience. A fix which has been run across the full test suite has been merged into main. If you rebase this _should_ resolve the issue

susnato commented:
Hmm it failed on another test :( but it was ok for the CLVP PR,  let me rebase and push one more time.

susnato committed:
very very small doc change

susnato commented:
Hi @amyeroberts, the CI is green now!

amyeroberts commented:
@susnato Amazing! Thanks for your work adding this and your patience whilst we sorted out the CI

Rocketknight1 commented:
Hi @susnato, congrats on the PR! Let me know if you're willing to tackle opening PRs to the model repos as well. I mentioned the recommended approach [here](https://github.com/huggingface/transformers/pull/26170#issuecomment-1787585303).

susnato commented:
Hi @Rocketknight1, yes I will open PR to both of the repos on the Hub this week.

Rocketknight1 commented:
Amazing, thank you!

susnato commented:
Hi @Rocketknight1, sorry I could not open the PRs this week, I will try to complete them ASAP.

susnato commented:
Hi @Rocketknight1, I have created both the PRs - [phi-1](https://huggingface.co/microsoft/phi-1/discussions/5) and [phi-1.5](https://huggingface.co/microsoft/phi-1_5/discussions/62).

Could you please review them?

liziniu commented:
Hi all,

It seems that the current code for Phi-1.5 does not support gradient checkpointing. Could you support this?

susnato commented:
Hi @liziniu, gradient checkpointing is already supported by `Phi`.

```python
from transformers import PhiForCausalLM

model = PhiForCausalLM.from_pretrained("susnato/phi-1_5_dev") # the weights are currently being transferred to microsoft/phi-1_5
print(model.model.gradient_checkpointing) # False

# enable Gradient Checkpointing
model.gradient_checkpointing_enable()

print(model.model.gradient_checkpointing) # this should be True now.
```

mohamedjihedriahi commented:
Hello folks

I've encountered an issue while attempting to load my fine-tuned model after the recent update. The error message I received is as follows:

"OSError: microsoft/phi-1_5 does not seem to contain a file named configuration_mixformer_sequential.py. Please check 'https://huggingface.co/microsoft/phi-1_5/main' for available files."

As someone relatively new to this, can anyone help me figure out how to tweak my code so I can load my fine-tuned model properly?
```
model = AutoModelForCausalLM.from_pretrained("nulltella/phi-1_5-finetuned-model", trust_remote_code=True, torch_dtype=torch.float32)
tokenizer = AutoTokenizer.from_pretrained("microsoft/phi-1_5", trust_remote_code=True)
```
Thanxxx !
#Help

susnato commented:
Hi @mohamedjihedriahi, as far as I can tell from the error message, your code is trying to find `configuration_mixformer_sequential.py` but that file was deleted in this [commit](https://huggingface.co/microsoft/phi-1_5/commit/8e9ebfb9bfe4da87d5c705d097fe884dc8e62d6f). So you are getting this error.


**Short Time Fix**
For a quick fix, you can try to convert your `nulltella/phi-1_5-finetuned-model` to have the same `config.json` as `microsoft/phi-1_5` according to [this commit](https://huggingface.co/microsoft/phi-1_5/commit/271c3397ab4e1f8f4e49868b1e8ba0be95363c88).

**Long Time Fix (Recommended)**
Since `phi` is now added to `transformers`, I would suggest you - 
1. upgrade to the latest `transformers`.
2. convert the previous weights according to the library model's weight structure. (this would require you to rearrange the keys of the weights).

This way, you won't have to care about any new changes to the model `phi` in the library. 


Let me know if you need any assistance/help in doing this.
